src/core/qgsmaplayer.h

```diff
@@ -256,9 +257,28 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     static QString extensionPropertyType( PropertyType type );
 
-    //! Returns the layer's unique ID, which is used to access this layer from QgsProject.
+    /**
+     * Returns the layer's unique ID, which is used to access this layer from QgsProject.
+     *
+     * \see setId()
+     * \see idChanged()
+     */
     QString id() const;
 
+    /**
+     * Sets the layer's \a id.
+     *
+     * \warning It is the caller's responsibility to ensure that the layer ID is unique in the desired context.
+     * Generally this method should not be called, and the auto generated ID should be used instead.
+     *
+     * \warning This method should not be called on layers which already belong to a QgsProject or QgsMapLayerStore.
```

Comment:
not saying this must be done, but potentially QgsProject or QgsMapLayerStore could call a QgsMapLayer::setSealedId() when they have acquired a layer, that would set a member variable mSealedId=true, and setId() could loudly complain then

Reply:
OK, I've implemented a variant of this!